src/hotspot/share/services/heapDumper.cpp

```diff
@@ -1446,18 +1790,152 @@ void HeapObjectDumper::do_object(oop o) {
   }
 }
 
+bool HeapObjectDumper::is_large(oop o) {
+  size_t size = 0;
+  if (o->is_instance()) {
+    InstanceKlass* ik = InstanceKlass::cast(o->klass());
+    size = DumperSupport::instance_size(ik);
```

Comment:
Getting the size of an instance can be surprisingly expensive for classes with many static fields, since we iterate over all static fields of the class. So I would avoid having to calculate the size twice (here and when we are actually dump it). Maybe here it would just be enough to use o->size() * 8 as an upper limit value of the real size in the heap dump. After all practically no non-array object will ever we that large.

Reply:
Good suggestion! I didn't realize that o->size() * 8 can be used as an upper limit.